pkg/scheduler/backend/queue/backoff_queue.go

```diff
@@ -0,0 +1,231 @@
+
+	return bq
+}
+
+func (bq *backoffQueue) podInitialBackoffDuration() time.Duration {
+	return bq.podInitialBackoff
+}
+
+func (bq *backoffQueue) podMaxBackoffDuration() time.Duration {
+	return bq.podMaxBackoff
+}
+
+// podsCompareBackoffCompleted is a less function of podBackoffQ and podErrorBackoffQ.
+func (bq *backoffQueue) podsCompareBackoffCompleted(pInfo1, pInfo2 *framework.QueuedPodInfo) bool {
+	bo1 := bq.getBackoffTime(pInfo1)
+	bo2 := bq.getBackoffTime(pInfo2)
+	return bo1.Before(bo2)
+}
+
+// shouldBeBackingoff returns true if a pod is still waiting for its backoff timer.
+// If this returns true, the pod should not be re-tried.
+func (bq *backoffQueue) shouldBeBackingoff(podInfo *framework.QueuedPodInfo) bool {
+	boTime := bq.getBackoffTime(podInfo)
+	return boTime.After(bq.clock.Now())
+}
+
+// getBackoffTime returns the time that podInfo completes backoff
+func (bq *backoffQueue) getBackoffTime(podInfo *framework.QueuedPodInfo) time.Time {
+	duration := bq.calculateBackoffDuration(podInfo)
+	backoffTime := podInfo.Timestamp.Add(duration)
+	return backoffTime
+}
+
+// calculateBackoffDuration is a helper function for calculating the backoffDuration
+// based on the number of attempts the pod has made.
+func (bq *backoffQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInfo) time.Duration {
+	if podInfo.Attempts == 0 {
+		// When the Pod hasn't experienced any scheduling attempts,
+		// they aren't obliged to get a backoff penalty at all.
+		return 0
+	}
+
+	duration := bq.podInitialBackoff
+	for i := 1; i < podInfo.Attempts; i++ {
+		// Use subtraction instead of addition or multiplication to avoid overflow.
+		if duration > bq.podMaxBackoff-duration {
+			return bq.podMaxBackoff
+		}
+		duration += duration
+	}
+	return duration
+}
+
+func (bq *backoffQueue) forEachBackoffCompletedWithQueue(logger klog.Logger, fn func(pInfo *framework.QueuedPodInfo), queue *heap.Heap[*framework.QueuedPodInfo]) {
+	for {
+		pInfo, ok := queue.Peek()
+		if !ok || pInfo == nil {
+			break
+		}
+		pod := pInfo.Pod
+		if bq.shouldBeBackingoff(pInfo) {
+			break
+		}
+		_, err := queue.Pop()
+		if err != nil {
+			logger.Error(err, "Unable to pop pod from backoff queue despite backoff completion", "pod", klog.KObj(pod))
+			break
+		}
+		fn(pInfo)
+	}
+}
+
+// forEachBackoffCompleted runs fn for all pods from podBackoffQ and podErrorBackoffQ that completed backoff.
+func (bq *backoffQueue) forEachBackoffCompleted(logger klog.Logger, fn func(pInfo *framework.QueuedPodInfo)) {
+	// Ensure both queues are called
+	bq.forEachBackoffCompletedWithQueue(logger, fn, bq.podBackoffQ)
+	bq.forEachBackoffCompletedWithQueue(logger, fn, bq.podErrorBackoffQ)
+}
+
+// add adds the pInfo to backoffQueue.
+// It also ensures that pInfo is not in both queues.
+func (bq *backoffQueue) add(logger klog.Logger, pInfo *framework.QueuedPodInfo) {
```

Comment:
nit: I prefer to name it `addOrUpdate` explicitly like the original interface. When you see `add` and `update` in the interface, you cannot tell what happens if the stuff already exists and then call `add` for it; whether you should `get` once to see whether stuff already exists or not, and then call either `add` or `update`. 

```suggestion
func (bq *backoffQueue) addOrUpdate(logger klog.Logger, pInfo *framework.QueuedPodInfo) {
```

Reply:
But this method should not be used to update anything. I am against sharing responsibility if possible. If there are `add` and `update` methods, it is understandable to me that `add` should be used only when adding a non-existent item.

Reply:
Yeah, we'd have a collision with the real 'update' function. Current naming seems aligned with other similar cases.